### ISSUE TITLE
new widgets system

### ISSUE THREAD

**Issue by marcelveldt** [Ignoble61/script.skinshortcuts#2]
I finally had the time to test out your work on the widget stuff. First off all: really great work. This is what I was looking for for a long time. I will also take a look at the templates stuff later but this already is very promising.

Is this ready to be merged to the main branch already ? Now there are a number of branches with altered stuff. It would be very, very nice if we could merge it all together in the main branch and start testing. I'm thinking about altering my Titan skin entirely to embrace this new widget (and maybe even the templates) stuff and provide a new beta version of my skin with a beta version of skinshortcuts to get some actual users to test it out.

So far, this new features are really things users have asked for:

- the speed improvements (already merged in main and working great)
- the "move up"  in the browse library nodes (I'm not sure where this was added, I suspect in the widgets branch of yours)
- the "edit action" stuff (work on progress on my side)
- the widget improvements.

As you may know I currently use skin strings to store the widget preference. I've incorporated the "just select shortcut" method for that to store what the users chooses from the library. My skin's background service will "translate" the normal path to a path to be used in the widget. In other words: it's a very messy workaround which I'd love to drop once this new widget stuff is ready. The only problem I have is how to migrate the current user preferences to the proper skinshortcuts implementation but I will figure something out, will be fine.

Just another question about the widgets. I'd like to have a button within the widget (on homescreen) that let's the user assign the widget directly. Now I already have this button there for some "predefined widgets"  that toggles between various states but I'd like to be able to just present the full dialog of choices the user has for the widget. Would that even be possible ? I'm trying really hard to get familiar with all the code to help you out, even while you where on dark for a little while I've spent days to scan through the code to fix some small issues users had but I'm not sure about this one. Maybe you can tell me from the top of your head if there's such an option in the code already. If not, it's a feature request :-)
But.. Like I said I want to help you our where I can with the code so if you could point me just a little in the right direction I'll try to add it in myself. I think it's a nice option that all skinners can benefit from.

So, I'm thinking about something like the "just select shortcuts" script entry, but this one will not be storing stuff to a skin string but directly in the mainmenu. Something like this perhaps:
RunScript(script.skinshortcuts,type=selectWidget&amp;menuId=300)

Thanks!

**Comment by Ignoble61** [comment]
Glad you like it :)

No, it's not ready to be merged into the main branch. A few issues remaining, aside from fixing up any little bugs (it's a lot of code, there WILL be little bugs introduced)...

* Not always returning correct mediaType for library nodes (working on fixing this right now, as it goes)
* Add-ons only return mediaType of Audio or Video, and even this isn't particularly reliable - if you know how to get the content for an addon from JSON, that would be very helpful ;)
* More add-ons need to be browsable (saw code from Nessus that defined dozens of extended info widgets - that script currently isn't browsable, but should be so they don't need to be defined by the skin, for example) - work on this has started, shouldn't take long to finish when I get back to it ;)
* Need to at least think about changing how skin widgets are defined so they can be better integrated into the groupings - if it happens it needs to happen at the same time as all this code landing

Sorry not sure what you mean in your last paragraph. Button within the button...? Full dialog of choices...?

**Comment by marcelveldt** [comment]
1) I have not yet seen issues with the mediatype but if you say so ;-)
2) Yes, I fact I do. I use that approach in my "smart shortcuts" feature from my skin. You can browse any content you like/want with json by using the Files.getDirectory method, even addons. Note that the addon should be a valid plugin which actually presents content, so no scripts (like extendedinfo).
3) Yep, I have that list of widgets too in my skin. Nice idea to incorporate that into the skinshortcuts. Will save skinners lot's of time.
4) Don't understand this one exactly ? You mean that any skinprovided widgets can be presented in the correct grouping/node within the node browser ? That would be great offcourse. On the other hand, the other possibiblity would be if the skinner can create it's own node/grouping to present the skin defined widgets. Offcourse the first option would be the best, that way the widgets show up in the correct grouping.
5) I've edited my post above, it was a typo. I meant to present a button within the homepage itself to select the widget.
6) Too bad that it is not yet ready. I guess I'll create a branch for my skin too where I start working on the widget changes and merge that back in later. The problem I have with branches is that you sometimes start losing the oversight of all changes. Not to speak of any merge conflicts. I'd rather just work on the main branch with all changes and just hold off with any puslishes to users until ready. For example. I hope you get what I mean :-)

**Comment by marcelveldt** [comment]
json_query_string = xbmc.executeJSONRPC('{"jsonrpc": "2.0", "method": "Files.GetDirectory", "params": { "directory": "plugin://plugin.video.netflixbmc/?mode=listSliderVideos&thumb&type=both&widget=true&url=slider_38", "media": "files", "properties": [ "title", "playcount", "plot", "file", "rating", "resume", "art", "streamdetails", "year", "mpaa", "runtime", "writer", "cast", "dateadded", "lastplayed", "tagline" ] }, "id": "1"}')
        json_result = json.loads(json_query_string)
        if json_result.has_key('result') and json_result['result'].has_key('files'):
            for item in json_result['result']['files']:

This example opens a path within a video addon and retrieves the listitems. You can base the content on the properties that are filled in. It will always be a best guess but you can get it pretty good. For example if it's an episode the listitem will have "season" and "episode" filled in. I think most part of this browsing idea is already in the script. Just one step further is to actually load the path and evaluate the listitem it returns.

please let me know if and where I can help.

**Comment by marcelveldt** [comment]
hmmm, wait... I think there's even a better way to solve this because each addon sets the content to a specific type. That is needed to let skins switch to the correct views. So it may not even be needed to evanulate the listitems itself. In the addon's code there is a line called "xbmcplugin.setcontent". that would set the content to a specific type. If only that preoperty could be returned somewhere in the json that would be great. I can look into it if you want, maybe even ask around with the Kodi devs on the forums.

**Comment by Ignoble61** [comment]
1) As of about 5 minutes ago, it shouldn't ;) (it was things like recently added tv shows/music videos - the .xml file defining the node doesn't actually specify the content type)
2) Already using the Files.getDirectory method to browse into them to allow a node to be selected as a shortcut :) The problem is getting the content type of the node (is it movies, tvshows, etc) - oh, and script.extendedinfo is also a plugin... ;) (and with the widgets branch should already be explorable under Program Add-ons, still needs work to show up under videos/music as appropriate)
edit - see you've updated your answer since I wrote that. Yup, that getting extra detail like season, episode and then taking a best-guess looks good :) (But a job for another night!)
3) That's the idea :)
4) The first idea is worth thinking about, and if it's to be implemented needs to be implemented at this stage :) The second is already doable, and may be enough if the amount of widgets skins provide can be cut significantly...
5) Not possible, but probably wouldn't take TOO much code refactoring to do. Two big jobs needed - and would take a fair bit of work:-
* Load available widgets in library.py, rather then gui.py
* Copy Just Select method, but changing it to call SelectShortcut as is done for widgets
* (Big job) Assign to the correct menu item (and I wouldn't have an idea how to start on this one...)
6) Down to personal preference, but I find branches much easier - just worry about one feature in one place, and only worry about bringing it all together when its actually working :)

**Comment by marcelveldt** [comment]
OK, thanks for your answers! I will create a branch for my skin too to incorporate the widget changes until this is merged to main as this involves with a huge refactor of my skin too ;-)

Just looked into the documentation for Files.GetDirectory:  The listitems should have a value for "type" which would be exactly what you need. I will do a small test and let you know.

As for the "just select widget" method. I think I get what you're saying here. If you want I could get a shot at fixing it if that saves you time.

**Comment by Ignoble61** [comment]
The type property invariable gives type="unknown" - in fact I don't think I've ever seen it return anything else! (And, even if it does return the correct value on rare occasions, it's too rare to be useful)

If you want to take on Just Select Widgets, that's great :)

**Comment by marcelveldt** [comment]
You're right, the type property does always return "unknown". I tried several ways but it's not returning the correct type. Only solution would be the "smart guess" based on the properties that are returned then. Maybe with a fallback to just "video", "audio" and "pictures" because we know what type of addon was launched. Do you want any help on this ? I can maybe think out the logic for the checks to determine the correct types ?

**Comment by Ignoble61** [comment]
The logic would be great. It needs some thinking about exactly how to implement, as currently the actual playable items from an add-on are completely ignored by the script. Plus, I need to think through various behaviours, such as what if content type is mixed. This could be what my grandson refers to as a "fun one"...!

**Comment by marcelveldt** [comment]
I think you can not cover every scenario. Maybe just "probe" 10 items of a list to determine the content. If it's all the same, set the content. If it's mixed or unknown then just set the type to mixed.
In fact I use mixed widgets all the time in my skin and I have the logic to determine the type in my skin itself and do not rely on what I know about the widget because that's now entirely out of my control. With this implementation in skinshortcuts however it may come a step closer of providing unified code in skins. Where do you want me to help you out ? In the code ? Or rather "on paper" by writing down the scenarios how to determine the content types. If the first, where's the code that browses the widget content located ?

BTW: just a little thing I saw... I think the label should be changed when the user browses for a widget from "select shortcut here" to "select widget here" or something like that in proper English ;-)

**Comment by Ignoble61** [comment]
That's why it needs thinking about ;)

Help out however you want to help out. Just getting the stuff "on paper" would be great. If you feel like playing with the code, that's just as great :) (I won't be touching any more code until tomorrow, myself).

The relevant parts of library.py:

* Line 1419 - files.getDirectory (needs extending to get media information to take best guess from)
* Line 1467 - where individual items from a plugin are processed (where the 'best guess' would be made for an individual item, in an `else:` block to the `if:` block here)
* Line 1501 onwards - where the widgetType (and other properties) are added to the shortcut that gets passed all the way back to the select widget function. (This will most likely need some playing about with to get right ;))

Nice spot on the string, I'll add it to my to-do list.

**Comment by marcelveldt** [comment]
I can take a look now if you want. Or are you already working something out ?

**Comment by Ignoble61** [comment]
I haven't looked at that at all yet. Are you planning to play with the code, or do the paper route?

(If it's the former, once your edit button commit is in and I've rebased the branch, I can give you commit rights if that's useful)

**Comment by marcelveldt** [comment]
Currently I'm playing with the code. I can either share that code with you later (on paper) or commit it. That's up to you ;-)

I think I have found a way to make it about 99% accurate based on guessing the content, but I have to test my theory

**Comment by Ignoble61** [comment]
If you've got the code I definitely want the code :) (especially if it's that accurate!) It's marginally easier for me in the long-run for you to commit it directly to the widgets branch, but a PR will work fine (just ever-so-slightly more work squashing a PR than a commit, and less chance of having merge conflicts), so I'll leave that as your choice.

**Comment by marcelveldt** [comment]
I'm working something out. Aready have detection for movies, tvshows, episodes and musicvideos. Now moving on to music items. I wil share my code later with you

**Comment by Ignoble61** [comment]
I've had this issue working its way through the back of my mind for over a week without a solution, so I'm genuinely excited to see what you have :)

(And, as far as I can see, this is the last big issue for the new widgets. Just a few niggles to sort out - and to discover, most likely, as I fully implement them into my own skin - and once this is sorted, it really shouldn't be long before these can be merged into the master :) )

**Comment by marcelveldt** [comment]
Well, good news... I'm testing now for about 20 minutes with some addons and different entrypoints and so far every match has been made ;-)
Do you want me to share the code I have so far for the video items ?

**Comment by Ignoble61** [comment]
I'm not doing any more coding tonight, so I'll leave it up to you if you want to share what you have so I can stare agog, or if you want to wait until it's all finished and I can look at implementing it (assuming you're working on it in a standalone way) :)

**Comment by marcelveldt** [comment]
I created a new method in library.py to do the lookup. You can just send he result of the json request to it and it will return the contenttype. Main listings are ignored, only actual content will have the guessing applied.

I'm not sure how to share this with you... I cn paste the contents here but I'm not sure if that would work properly. pastebin maybe ?
Other solution would be if I clone your repo. Not sure how to clone a branch though...

**Comment by Ignoble61** [comment]
Not sure myself, so let me just add you to my branch and it should* show in whatever git client you're using (normally after a restart and a sync) as a branch you can select and then work on, and commit directly to it.

If that doesn't work, getting a diff (not sure how you do this on Window) would work great too. Worst comes to the worst, just pass the relevant methods here :)

* insert disclaimer here about how these things never work as you expect

**Comment by marcelveldt** [comment]
hehe, I will try to commit the changes so you can look at it yourself and we can work from there.
It's basically a whole new method so it cannot harm I think. Also in the explorer method it will now only print the found itemType. It's not yet assigned.

BTW: I just found out that the detection also works for custom library nodes.

**Comment by marcelveldt** [comment]
OK, fongers crossed. I've just committed the changes. Please take a look. In the meanwhile I am going to search for an audio addon to integrate the audio items detection.

**Comment by Ignoble61** [comment]
Assigning it will take some playing with - the current code to assign all of the widget values in the explorer function has been quickly hacked on top of the existing code, and as such is a complete mess and needs cleaning up when I have time :) Just getting the value is more than perfect for the moment! (And I'd almost rather it doesn't return the value atm, rather than further hack the code. Better me spending the time to do the code right)

Custom library nodes SHOULD have the type assigned in the relevant xml files (and therefore SHOULD be returned correctly as is), but if this provides a backup for when the user hasn't done what's expected of them, well that's even better :)

**Comment by marcelveldt** [comment]
Give it a look/try and let me know. There's still some debugging "print" lines in there. They can be removed offcourse.

**Comment by Ignoble61** [comment]
A quick look and I love the concept :) I would never have come up with this!

My initial thoughts is it will probably need a small refactor of the explorer() to function correctly (or, at the very least, optimally), but that's very doable - most notably, the explorer function works on individual items, not the whole lot - which is most notable when it comes to add-ons, for which this is most directly relevant. If at all possible (it may not be) either both operate on the whole list, or both operate on individual items, otherwise everything gets processed twice.

One thought was it would be interesting to know the speed differences between checking 5 items and checking, for example, 50 - my suspicion is it wouldn't actually make much difference from a speed perspective (the big sloth here is getting the json - most functions parsing it run for fractions of a second), though as your comments point out, 'Next' items could impact this giving an accurate result. It may be a percentage is best for deciding the content (e.g. if > 90% is one content type, that's the content type)

I'll actually checkout the changes tomorrow and have a play with it, but so far so great :)

**Comment by marcelveldt** [comment]
Glad I could be at any help of your amazing work. Refactor/play with it as you find is best. It's your game ;-)

I will now look at the music stuff. Been wondering how to get some actual music addons to play with. Guess I'll add some music to my test Plex instance to play with...
I think the music section will be as much of the same.

**Comment by Ignoble61** [comment]
I'm always very wary of messing with other peoples code (people can be very territorial of their code!), especially when it's such a fundamentally great concept at this. But, as I've got your permission and because it's the last missing piece for these new widgets, I'll fully enjoy playing with this code and its integration tomorrow :)

Thanks so much for this - you've just made the add-ons useful as widgets within this new code :+1:

**Comment by marcelveldt** [comment]
Not worry about it, you are the father of this amazing script. I just help you around where I can. Do with the code what you think is best. Also, I'm not the best coder in the world. I just have the logic in my head how to create stuff and code it with google ;-) I guess that's why at work I'm the functional designer and not the coder :-)

I just committed the detection for audio too btw.

**Comment by Ignoble61** [comment]
> I guess that's why at work I'm the functional designer and not the coder :-)

Don't sell yourself short - 90% of programming is working out HOW to solve the problem. Only 10% is actually writing the code (which you're better at than you give yourself credit), and you've proven yourself time and again as a problem solver!

**Comment by marcelveldt** [comment]
Thanks :-) Enjoy your evening.

**Comment by Ignoble61** [comment]
> Just another question about the widgets. I'd like to have a button within the widget (on homescreen) that let's the user assign the widget directly. Now I already have this button there for some "predefined widgets" that toggles between various states but I'd like to be able to just present the full dialog of choices the user has for the widget. Would that even be possible ? I'm trying really hard to get familiar with all the code to help you out, even while you where on dark for a little while I've spent days to scan through the code to fix some small issues users had but I'm not sure about this one. Maybe you can tell me from the top of your head if there's such an option in the code already. If not, it's a feature request :-)
> But.. Like I said I want to help you our where I can with the code so if you could point me just a little in the right direction I'll try to add it in myself. I think it's a nice option that all skinners can benefit from.
> 
> So, I'm thinking about something like the "just select shortcuts" script entry, but this one will not be storing stuff to a skin string but directly in the mainmenu. Something like this perhaps:
> RunScript(script.skinshortcuts,type=selectWidget&menuId=300)

Had some time to think through this, and think I understand what you're wanting to do. If I'm understanding correctly there are two issues you're going to have to work around:-

* Which menu item you're editing the widgets for. Normally, the script uses the labelID to identify individual menu items. In this case, each menu item has an id integer - if you can get this (honestly, no idea if this is actually accessible, I only remember it was requested many ages ago) then it may make things easier, otherwise the labelID should be able to be used as the actual main menu items aren't changing (which is when the labelID's may change). Either way, you need to identify - in a way the script can understand - which main menu item is currently selected
* Update the various widget properties for the menu item. There's currently no code to do this, but it shouldn't be too much of a stretch to load the menu items, then identify the correct one (as per the previous item), then just update the relevant properties and save them. If you're only doing this for main menu items in a menu fully-managed by the script, this shouldn't be too difficult, but to do it well may require refactoring some code from gui.py.

Where you may find issues is in the user experience - to actually change the values in script-skinshortcuts-includes.xml, the whole menu needs to be rebuilt. Not a problem, but the skin has to be reloaded as part of that process, so the highlighted menu item may change (for example), which may not lead to great UX.

If you need help doing this, let me know.

**Comment by Ignoble61** [comment]
widgetType is now being returned for addons. Thanks again for the code :)

**Comment by marcelveldt** [comment]
Great! I will test it tomorrow once I'm back home again.

**Comment by Ignoble61** [comment]
Thanks for all your help testing (and coding) this :)
